include/kframework/foundry.md

```diff
@@ -657,6 +657,22 @@ Otherwise, throw an error for any other call to the Foundry contract.
       [owise]
 ```
 
+#### `sign` - Signs a digest with private key
+
+```
+function sign(uint256 privateKey, bytes32 digest) external returns (uint8 v, bytes32 r, bytes32 s);
+```
+
+`foundry.call.sign` will match when the `sign` cheat code function is called.
+This rule then takes from the `privateKey` to sign using `#range(ARGS,0,32)` and the `digest` to be signed using `#range(ARGS, 32, 32)`. To perform the signature we use the `ECDSASign ( String, String )` function (from KEVM). This function receives as arguments 2 strings: the data to be signed and the private key, therefore we use `#unparseByteStack` to convert the bytearrays with the `privateKey` and `digest` into strings. The `ECDSASign` function returns the signed data in [r,s,v] form, which we convert to a bytearray using `#parseByteStack`.
+
+```{.k .bytes}
+    rule [foundry.call.sign]:
+         <k> #call_foundry SELECTOR ARGS => . ... </k>
+         <output> _ => #parseByteStack(ECDSASign(#unparseByteStack(#range(ARGS, 32, 32)),#unparseByteStack(#range(ARGS,0,32)))) </output>
```

Comment:
Hmmmm, this direct call of `ECDSASign` probably is not a good idea, because it will try to evaluate as a hooked symbol, which there may not be implementations for.

Maybe better to make a wrapper like this:

```
syntax ByteArray ::= #sign ( ByteArray , ByteArray ) [function]
rule #sign(BA1, BA2) => #parseByteStack(ECDSASign(#unparseByteStack(BA1), #unparseByteStack(BA2))) [concrete]
```

Then use `#sign` in this rule.

Currently the only test of this only feeds concrete values into `ECDSASign`, which is likely why you haven't see the backend blow up on a symbolic inputs to this yet. So this proposed modification will be fine for your test-suite (because it only exercises the concrete case), but will keep the backend from blowing up in the symbolic case (which you maybe should consider adding a test for).

Reply:
I added a fuzzing test and indeed the rule doesn't work anymore. I will address your suggestions and see if it fixes.